### PR TITLE
Do IMAP auth over direct socket connection

### LINKED ISSUE #13: OC_User_IMAP supporting only TLSv1
### Affected apps
user_external
### Steps to reproduce
1.
    Use OC_User_IMAP for authentication
    {imapserver:143/imap/tls/readonly}
2. Disable TLSv1 on the IMAP server remaining TLSv1.1 and TLSv1.2

See closed issue for Nectcloud nextcloud/server#10328

### LINKED ISSUE #44: PHP IMAP extension deprecated
### Steps to reproduce
1. Search for php7-imap in the repo's of current distro's like openSUSE Tumbleweed
2. Lookup php-imap extension status in the formal PHP extension list

### Expected behaviour
The native PHP IMAP extension appears to be deprecated.

### Actual behaviour
The php-imap extension appears to have been abandoned by its maintainers and distro's like openSUSE no longer ship it.
https://wiki.php.net/todo/extensions#imap
https://bugzilla.opensuse.org/show_bug.cgi?id=1089061#c18

Roundcube and other mail apps appear to have moved to a custom PHP IMAP protocol implementation using the generic sockets extension.

Would it be possible to consider replacing this unmaintained dependency to a similar solution?
https://github.com/roundcube/roundcubemail/blob/82974289b98d2d00251e55f1719c4522598d7ef5/program/lib/Roundcube/rcube_imap_generic.php

### Affected Authentication backend
IMAP

### Server configuration

**Operating system**:
Several linux distro's i.e. openSUSE

**PHP version:**
>7

### PR DESCRIPTION
**Pull request by violoncelloCH**
fixes #44 
fixes #13

Like proposed in #44 we took the implementation of roundcube but stripped away everything that's not needed for the authentication.

as we're not using the php-imap (imap_open) library, this dependency isn't needed any more, however we need some different parameters to be set, so the configuration is different and needs to be changed on running instances (see https://github.com/nextcloud/user_external/blob/fe30fbf5156d221449180f7d576a97aba522f1ee/README.md vs. the current one)

thank you very much @ChristophWurst for your help!

cc @nextcloud/user_external @hcderaad @patschi @andryyy @lsbbs

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -56,38 +56,29 @@ needs to be activated.
...
         ),
     ),
 
-This connects to the IMAP server on IP `127.0.0.1`, in readonly mode.
+This connects to the IMAP server on IP `127.0.0.1`.
+The default port is 143; however if you want to restrict the domain, you need to specify the port and set sslmode to `null`.
 If a domain name (e.g. example.com) is specified, then this makes sure that
 only users from this domain will be allowed to login. After successfull
 login the domain part will be striped and the rest used as username in
 NextCloud. e.g. 'username@example.com' will be 'username' in NextCloud.
```
Comment (MariusBluem) [inline README.md]: `NextCloud `-> `Nextcloud`

Comment (violoncelloCH) [reply README.md]: good point! :sweat_smile:

**Comment by philtrick-twm** [comment]
Just a bit of information, I needed to use the following in the config.php: (I needed to explicitly use the mail server name, rather than 127.0.0.1, and I had to put single quotes around the ssl parameter, otherwise nextcloud had an error)

'user_backends' => array(
    array(
        'class' => 'OC_User_IMAP',
        'arguments' => array(
            'mail.example.com', 993, 'ssl', 'example.com'
        ),
    ),
),

**Comment by violoncelloCH** [comment]
@philtrick-twm yes, thank you, see #52 and #53

**Pull request merged**